src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/common/event_parser.h"  // For FrameBase.
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mqtt {
+
+// The protocol specification : https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.pdf
+// This supports MQTTv5
+
+struct Message: public FrameBase {
+    message_type_t type = message_type_t::kUnknown;
+
+    std::string control_packet_type = "UNKNOWN";
+
+    bool dup;
+    bool retain;
+
+    std::map<std::string, uint32_t> header_fields;
+    std::map<std::string, std::string> properties, payload;
+
+    template<typename KeyType, typename ValueType>
+    static std::string MapToString(const std::map<KeyType, ValueType>& inputMap) {
+        std::string result = "{";
+        for (const auto& entry : inputMap) {
+            result += entry.first + ": ";
+            if constexpr (std::is_same_v<ValueType, uint32_t>) {
+                result += std::to_string(entry.second);
+            } else if constexpr (std::is_same_v<ValueType, std::string>) {
+                result += entry.second;
+            }
+            result += ", ";
+        }
+        if (!inputMap.empty()) {
+            result = result.substr(0, result.size() - 2); // Remove the trailing ", "
+        }
+        result += "}";
+        return result;
+    }
```

Comment:
Could this be replaced with the [ToJSONString function?](https://github.com/pixie-io/pixie/blob/62b8080699371b0bcd83e2c3ddda265d22d0eb40/src/common/json/json.h#L119) It seems like this is similar to json encoding and we should use `rapidjson` rather than building it by hand.